Added decoding pipeline file

Benzillaist opened the pull request:
The evaluate_decoder method should be the only public facing method as the rest should only need to be called by evaluate_classical_decoder.

- Created 2 versions of evaluate_classical_decoder. One of them should work, and it takes in a decoder object. The type of the object detemines which decoder is called later on in the function. The decoding is done with a Pauli frame simulation based upon code written by Anthony for evaluating the belief propagation decoder.

- Tried to test my methods, but I've been having a problem with running standard Stabilizer type methods on Stabilizer objects that I created from a Hx and Hz matrix. Pointers on how to fix this would be appreciated!

- Included a CSS class below which includes some methods I use in the decoder, specifically the parity_checks method. In the future this will be added to QuantumClifford.

Benzillaist committed:
Added decoding pipeline file

The evaluate_decoder method should be the only public facing method as the rest should only need to be called by evaluate_classical_decoder.

I have created 2 versions of evaluate_classical_decoder. One of them should work, and it takes in a decoder object. The type of the object detemines which decoder is called later on in the function. The decoding is done with a Pauli frame simulation based upon code written by Anthony for evaluating the belief propagation decoder.

I have tried to test my methods, but I've been having a problem with running standard Stabilizer type methods on Stabilizer objects that I created from a Hx and Hz matrix. Pointers on how to fix this would be appreciated!

I have included a CSS class below which includes some methods I use in the decoder.

Krastanov commented on src/LDPCDecoders.jl:
```diff
@@ -6,6 +6,7 @@ using Random
 using SparseArrays
 using Statistics
 
+using QuantumClifford
```
We should not depend on QuantumClifford, at least for now

However, we can totally depend on it in tests

Krastanov commented on src/decoder_pipeline.jl:
```diff
@@ -0,0 +1,246 @@
+
+# Added for running above methods, should not be included in final code(?), implemented inside of QuantumClifford
+"""An arbitrary CSS error correcting code defined by its X and Z checks."""
+struct CSS <: AbstractECC
```
this should be part of QuantumClifford, not LDPCDecoders

Krastanov commented on src/decoder_pipeline.jl:
```diff
@@ -0,0 +1,246 @@
+    row_x = syndrome_sample[1:d.numchecks_X]
+    row_z = syndrome_sample[d.numchecks_X+1:d.numchecks_X+d.numchecks_Z]
+
+    KguessX, success = syndrome_decode(sparse(d.Cx), sparse(d.Cx'), d.row_x, d.max_iters, d.channel_probs, d.b2c_X, d.c2b_X, d.log_probabs, Base.copy(d.err))
```
If these sparse matrices are necessary, these should be created in advance, not each time a decoder is called. These can be, for instance, additional properties of the struct.

Krastanov commented on src/decoder_pipeline.jl:
```diff
@@ -0,0 +1,246 @@
+
+"""Type for evaluating the table decoder"""
+struct TableDecoder <: AbstractSyndromeDecoder
+    H
+    faults_matrix
+    n
+    s
```
Let's add some documentation inline for these structs. Something like this

```suggestion
    "the stabilizer tableau defining the code"
    H
    "the faults matrix that converts a physical error vector into corresponding logical error vector"
    faults_matrix
    n
    s
```

Krastanov commented on src/decoder_pipeline.jl:
```diff
@@ -0,0 +1,246 @@
+
+"""Type for evaluating the belief propagation decoder"""
+struct BeliefPropDecoder <: AbstractSyndromeDecoder
+    H
+    faults_matrix
+    n
+    s
+    k
+    log_probabs
+    channel_probs
+    numchecks_X
+    b2c_X
+    c2b_X
+    numchecks_Z
+    b2c_Z
+    c2b_Z
+    err
```
documenting these as mentioned above

Krastanov reviewed (commented):
- it is difficult to decide how to take care of interdependencies between LDPCDecoders and QuantumClifford. For the moment let's use the following structure:
  - QuantumClifford will start depending on LDPCDecoders
  - LDPCDecoders thus can not contain anything imported from QuantumClifford
  - LDPCDecoders should contain the API for classical decoders. Currently, that does not really exist, only an API for quantum decoders is implemented in this PR.
  - Let's move the quantum decoders API to QuantumClifford
  - No need to worry about creating a classical decoder API yet, it probably is not a good use of your limited time
  - If one day there is a classical decoder API in this package, the quantum decoder API in QuantumClifford should depend on the classical one, not the other way around

Benzillaist commented:
Addressed the above comments and moved the code to QuantumClifford.jl
[New pull request](https://github.com/QuantumSavory/QuantumClifford.jl/pull/209)
